### PR TITLE
login mvvm combine

### PR DESCRIPTION
**Pull request by Alexdive**

### CODE REVIEW

File: LogIn/LoginViewController.swift [hunk 1]
```diff
@@ -10,32 +10,54 @@ import Combine
 
 class LoginViewController: UIViewController {
```
Comment (0111b) [inline LogIn/LoginViewController.swift]:
```suggestion
final class LoginViewController: UIViewController {
```

File: LogIn/UIExtensions.swift [hunk 2]
```diff
@@ -5,23 +5,31 @@
 //  Created by Aleksei Permiakov on 07.12.2021.
 //
 
+
 import UIKit
 
 extension UITextField {
     func loginTextStyle() {
```
Comment (0111b) [inline LogIn/UIExtensions.swift]: maybe something like `applyLoginTextStyle`, because meaning is not clear

File: LogIn/UIExtensions.swift [hunk 3]
```diff
@@ -5,23 +5,31 @@
...
         self.backgroundColor = .systemGray6
```
Comment (0111b) [inline LogIn/UIExtensions.swift]: Also it make sense to make this function private and move to the login view

File: LogIn/LoginViewController.swift [hunk 4]
```diff
@@ -10,32 +10,54 @@ import Combine
...
     
+    var model: LoginViewModelType
+    
+    private let email = PassthroughSubject<String, Never>()
```
Comment (0111b) [inline LogIn/LoginViewController.swift]: To remove boilerplate and copy-paste, its nice to write extension to the TextField with custom publisher.

Comment (0111b) [reply LogIn/LoginViewController.swift]: Check out https://github.com/CombineCommunity/CombineCocoa/blob/main/Sources/CombineCocoa/Controls/UITextField%2BCombine.swift as an example reference

File: LogIn/LoginViewController.swift [hunk 5]
```diff
@@ -45,48 +67,89 @@ class LoginViewController: UIViewController {
...
     override func viewDidLoad() {
         super.viewDidLoad()
         
         setupViews()
+        bind(to: model)
+    }
+    
+    @objc
+    private func emailTextEntered() {
+        if let text = emailTextField.text {
```
Comment (0111b) [inline LogIn/LoginViewController.swift]: nil checking and other transformations  must be done in the viewModel

File: LogIn/LoginViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    let emailFormatMessage: String
+    let isEnabled: Bool
+}
+
+protocol LoginViewModelType {
+    var outputPublisher: Published<LoginViewModelOutput>.Publisher { get }
+    func transform(input: LoginViewModelInput)
+}
+
+class LoginViewModel: LoginViewModelType {
```
Comment (0111b) [inline LogIn/LoginViewModel.swift]: Why it is a `class` and why it is not `final`?

File: LogIn/LoginViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    
+    @Published private(set) var output: LoginViewModelOutput = .init(emailTint: .systemGray2,
```
Comment (0111b) [inline LogIn/LoginViewModel.swift]: You do not need this wrapper here, lets discuss.

File: LogIn/LoginViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        var emailTint: UIColor = .systemGray2
+        var passwTint: UIColor = .systemGray2
+        var passwAgainTint: UIColor = .systemGray2
+        var emailFormatMessage: String = ""
+        var isEnabled: Bool = false
+        
+        var isValid = false
+        
+        input.email
+            .debounce(for: 1, scheduler: RunLoop.main)
```
Comment (0111b) [inline LogIn/LoginViewModel.swift]: too long =)

File: LogIn/LoginViewModel.swift [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    
+    var outputPublisher: Published<LoginViewModelOutput>.Publisher { $output }
+    
+    private var subscriptions = Set<AnyCancellable>()
+    
+    func transform(input: LoginViewModelInput) {
+        subscriptions.forEach { $0.cancel() }
+        subscriptions.removeAll()
+        
+        var emailTint: UIColor = .systemGray2
```
Comment (0111b) [inline LogIn/LoginViewModel.swift]: Lets discuss the flow

**Review by 0111b — CHANGES_REQUESTED** [review]

File: LogIn/LoginViewController.swift (hunk 1 above)
Comment (Alexdive) [reply LogIn/LoginViewController.swift]: done

File: LogIn/UIExtensions.swift (hunk 2 above)
Comment (Alexdive) [reply LogIn/UIExtensions.swift]: done

File: LogIn/UIExtensions.swift (hunk 3 above)
Comment (Alexdive) [reply LogIn/UIExtensions.swift]: make sense. Done

File: LogIn/LoginViewController.swift (hunk 5 above)
Comment (Alexdive) [reply LogIn/LoginViewController.swift]: done

File: LogIn/LoginViewModel.swift (hunk 6 above)
Comment (Alexdive) [reply LogIn/LoginViewModel.swift]: made it final. Class to be able to use @Published wrapper

File: LogIn/LoginViewModel.swift (hunk 8 above)
Comment (Alexdive) [reply LogIn/LoginViewModel.swift]: well, its subjective. I don't like when app screaming on me right away

File: LogIn/LoginViewController.swift (hunk 4 above)
Comment (Alexdive) [reply LogIn/LoginViewController.swift]: good one!
Done

File: LogIn/LoginViewModel.swift (hunk 9 above)
Comment (Alexdive) [reply LogIn/LoginViewModel.swift]: adjusted flow

**Pull request merged**